scripts/support/zephyr_flash_debug.py

```diff
@@ -38,105 +40,236 @@ def get_env_bool_or(env_var, default_value):
         return default_value
 
 
+def get_env_strip_or(env_var, to_strip, default_value):
+    value = os.environ.get(env_var, None)
+    if value is not None:
+        return value.strip(to_strip)
+    else:
+        return default_value
+
+
+def quote_sh_list(cmd):
+    '''Transform a command from list into shell string form.'''
+    fmt = ' '.join('{}' for _ in cmd)
+    args = [shlex.quote(s) for s in cmd]
+    return fmt.format(*args)
+
+
 def check_call(cmd, debug):
     if debug:
-        print(' '.join(cmd))
+        print(quote_sh_list(cmd))
     subprocess.check_call(cmd)
 
 
 def check_output(cmd, debug):
     if debug:
-        print(' '.join(cmd))
+        print(quote_sh_list(cmd))
     return subprocess.check_output(cmd)
 
 
-class ZephyrBinaryFlasher(abc.ABC):
-    '''Abstract superclass for flasher objects.'''
+def popen_ignore_int(cmd, debug):
+    '''Spawn a child command, ensuring it ignores SIGINT.
+
+    The returned subprocess.Popen object must be manually terminated.'''
+    cflags = 0
+    preexec = None
+    system = platform.system()
+
+    if system == 'Windows':
+        cflags |= subprocess.CREATE_NEW_PROCESS_GROUP
+    elif system in {'Linux', 'Darwin'}:
+        preexec = os.setsid
+
+    if debug:
+        print(quote_sh_list(cmd))
+
+    return subprocess.Popen(cmd, creationflags=cflags, preexec_fn=preexec)
+
+
+MAX_PORT = 49151
+
+
+class NetworkPortHelper:
+    '''Helper class for dealing with local IP network ports.'''
+
+    def get_unused_ports(self, starting_from):
+        '''Find unused network ports, starting at given values.
+
+        starting_from is an iterable of ports the caller would like to use.
+
+        The return value is an iterable of ports, in the same order, using
+        the given values if they were unused, or the next sequentially
+        available unused port otherwise.
+
+        Ports may be bound between this call's check and actual usage, so
+        callers still need to handle errors involving returned ports.'''
+        start = list(starting_from)
+        used = self._used_now()
+        ret = []
+
+        for desired in start:
+            port = desired
+            while port in used:
+                port += 1
+                if port > MAX_PORT:
+                    msg = "ports above {} are in use"
+                    raise ValueError(msg.format(desired))
+            used.add(port)
+            ret.append(port)
+
+        return ret
+
+    def _used_now(self):
+        handlers = {
+            'Windows': self._used_now_windows,
+            'Linux': self._used_now_linux,
+            'Darwin': self._used_now_darwin,
+        }
+        handler = handlers[platform.system()]
+        return handler()
+
+    def _used_now_windows(self):
+        cmd = ['netstat', '-a', '-n', '-p', 'tcp']
+        return self._parser_windows(cmd)
+
+    def _used_now_linux(self):
+        cmd = ['netstat', '-a', '-n', '-t']
```

Comment:
`netstat` has been deprecated in Linux; my system at home doesn't even have it installed.  The `ss` utility from `iproute2` is the replacement. It might be the case both need to be tried.

Reply:
Hrm, OK. Use of netstat matches the existing shell script, which I'm trying to replicate fairly closely in this series (except of course where it needs to be changed for non-Linux targets).

Would you be OK with me fixing this to use ss in a follow-up? I can do it now if that's better, but it feels like the wrong series to do it in.

Reply:
It can be done in a follow up patch, no problems.

Reply:
Thanks, I'll queue this up in the next series:

https://github.com/mbolivar/zephyr/commit/57e096f165d6c59affb42d719d69bb71fb51dc0c